### PR TITLE
Metrics implementation for MasterDataIndexerJob

### PR DESCRIPTION
**Pull request by SurabhiAngadi**
Changes -

1. Updated job to work with service account.
2. Job compatible for multiple cloud storage provider.
3. Added metrics for the job.
4. Sample metric event ->
{
"eid": "METRIC",
"ets": 1700474870038,
"mid": "b7e57824-baa9-4eb2-b452-00b1c74abd84",
"actor": {
"id": "MasterDataProcessorIndexerJob",
"type": "SYSTEM"
},
"context": {
"env": "local",
"pdata": {
"id": "DataProducts",
"pid": "MasterDataProcessorIndexerJob",
"ver": "1.0.0"
}
},
"object": {
"id": "telemetry-content-data",
"type": "Dataset",
"ver": "1.0.0"
},
"edata": {
"metric": {
"metric_name": 1,
"total_time_taken": 34977,
"total_events_processed": 10000, ...etc.
},
"labels": [
{
"key": "job",
"value": "MasterDataIndexer"
},
{
"key": "datasetId",
"value": "telemetry-content-data"
},
{
"key": "cloud",
"value": "aws"
}
]
}
}

Test scenarios to be performed -

1. Job should filter dataset based on indexData column and check if there are more than one datasource associated with each dataset. If so, return early.
2. Job should create data file and upload to respective cloud storage provider.
3. Submit ingestion spec to druid for processing.
4. Update datasourceRef if needed.
5. Job should generate metrics for both success and failure scenarios.

### CODE REVIEW

File: data-products/src/main/scala/org/sunbird/obsrv/dataproducts/MasterDataProcessorIndexer.scala [hunk 1]
```diff
@@ -2,129 +2,131 @@ package org.sunbird.obsrv.dataproducts
...
-  private def deleteDataSource(datasourceRef: String): Unit = {
-    // TODO: Handle success and failure responses properly
-    val response = Unirest.delete(config.getString("druid.datasource.delete.url") + datasourceRef)
-      .header("Content-Type", "application/json")
-      .asJson()
-    response.ifFailure(_ => throw new Exception("Exception while deleting datasource" + datasourceRef))
+  private def getDatasets(): List[Dataset] = {
+    val datasets = DatasetRegistry.getAllDatasets("master-dataset")
+    datasets.filter(dataset => {
+      logger.debug("Checking dataset status for id - " + dataset.id)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/obsrv/dataproducts/MasterDataProcessorIndexer.scala:90]: Remove debug statement

File: data-products/src/main/scala/org/sunbird/obsrv/dataproducts/helper/BaseMetricHelper.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  private def getObject(datasetId: String) = {
+    MetricObject(id = datasetId, `type` = "Dataset", ver = "1.0.0")
+  }
+
+  def generate(datasetId: String, edata: Edata): Unit = {
+    val `object` = getObject(datasetId)
+    val actor = Actor(id = "MasterDataProcessorIndexerJob", `type` = "SYSTEM")
+    val pdata = Pdata(id = "DataProducts", pid = "MasterDataProcessorIndexerJob", ver = "1.0.0")
+    val context = Context(env = config.getString("env"), pdata = pdata)
+    val metric = JobMetric(ets = new DateTime(DateTimeZone.UTC).getMillis, actor = actor, context = context, `object` = `object`, edata = edata)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/obsrv/dataproducts/helper/BaseMetricHelper.scala:38]: System.currentMillis() is enough. Not need to create new datetime objects unecessarily

File: data-products/src/main/scala/org/sunbird/obsrv/dataproducts/util/CommonUtil.scala [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    val conf = new SparkConf().setAppName(appName)
+    val master = getSafeConfigString(config, "spark.master")
+    if (master.isEmpty) {
+      logger.info("Master not found. Setting it to local[*]")
+      conf.setMaster("local[*]")
+    }
+    SparkSession.builder().appName(appName).config(conf).getOrCreate()
+  }
+  // $SCOVERAGE-ON$
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/obsrv/dataproducts/util/CommonUtil.scala:44]: Code can be covered

File: dataset-registry/src/main/scala/org/sunbird/obsrv/model/DatasetModels.scala [hunk 4]
```diff
@@ -93,5 +92,5 @@ class DatasetStatusType extends TypeReference[DatasetStatus.type]
 
 object DatasetStatus extends Enumeration {
   type DatasetStatus = Value
-  val Draft, Publish, Live, Retired, Purged = Value
+  val Draft, Publish, Live, Retired, Purged, ACTIVE = Value
```
Comment (SanthoshVasabhaktula) [inline dataset-registry/src/main/scala/org/sunbird/obsrv/model/DatasetModels.scala:95]: Revert the change. There is no status called ACTIVE. It is Live

File: framework/src/main/scala/org/sunbird/obsrv/core/util/PostgresConnect.scala [hunk 5]
```diff
@@ -72,6 +72,7 @@ class PostgresConnect(config: PostgresConnectionConfig) {
   }
 
   def executeQuery(query:String):ResultSet = statement.executeQuery(query)
+
```
Comment (SanthoshVasabhaktula) [inline framework/src/main/scala/org/sunbird/obsrv/core/util/PostgresConnect.scala:75]: Reset the file

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

File: data-products/src/main/scala/org/sunbird/obsrv/dataproducts/util/StorageUtil.scala [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  def getPaths(datasource: DataSource, config: Config): Paths = {
+    val dt = new DateTime(DateTimeZone.UTC).withTimeAtStartOfDay()
+    val timestamp = dt.getMillis
+    val date = dayPeriodFormat.print(dt)
+    val provider = providerFormat(config.getString("cloudStorage.provider"))
+    val cloudPrefix = if (config.getString("cloudStorage.provider") == "azure") {
+      provider.sparkURIFormat + config.getString("azure_cloud_prefix")
+    } else {
+      provider.sparkURIFormat + config.getString("cloud_prefix")
+    }
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/obsrv/dataproducts/util/StorageUtil.scala:42]: Why is there a check specific for azure?

File: data-products/src/main/scala/org/sunbird/obsrv/dataproducts/util/StorageUtil.scala [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    val outputFilePath = cloudPrefix + pathSuffix
+    Paths(datasourceRef, ingestionPath, outputFilePath, timestamp)
+  }
+
+  //This method provides appropriate input source spec depending on the cloud storage provider
+  def inputSourceSpecProvider(filePath: String, config: Config): String = {
+    val provider = StorageUtil.providerFormat(config.getString("cloudStorage.provider"))
+    val inputSourceSpec = if (provider.ingestionSourceType == "local")
+      config.getString("local_input_source_spec").replace("FILE_PATH", filePath)
+    else config.getString("cloud_input_source_spec").replace("FILE_PATH", filePath)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/obsrv/dataproducts/util/StorageUtil.scala:55]: Incorrect equals comparison
Remove the local file check

File: data-products/src/main/resources/masterdata-indexer.conf [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+ topicName = ""${env}".spark.stats"
+}
+
+kafka {
+    bootstrap.servers = "localhost:9092"
+}
+
+#inputSourceSpec
+local_input_source_spec="{\"spec\":{\"ioConfig\":{\"type\":\"index_parallel\",\"inputSource\":{\"type\":\"local\",\"baseDir\":\"FILE_PATH\",\"filter\":\"**json.gz\"}}}}"
+cloud_input_source_spec="{\"spec\":{\"ioConfig\":{\"type\":\"index_parallel\",\"inputSource\":{\"type\":\"${cloudStorage.provider}\",\"objectGlob\":\"**.json.gz\",\"prefixes\":[\"FILE_PATH\"]}}}}"
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/resources/masterdata-indexer.conf:24]: Remove variable specific to local

**Review by SanthoshVasabhaktula — CHANGES_REQUESTED** [review]

**Review by manjudr — APPROVED** [review]

**Comment by manjudr** [comment]
The changes are looks fine.

**Pull request merged**